File: aoc-2023/src/main/java/com/capital7software/aoc/aoc2023/days/Day01.java
package com.capital7software.aoc.aoc2023.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.string.Trebuchet;
import java.time.Instant;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * --- Day 1: Trebuchet?! ---<br><br>
 * Something is wrong with global snow production, and you've been selected to take a look.
 * The Elves have even given you a map; on it, they've used stars to mark the top fifty locations
 * that are likely to be having problems.
 *
 * <p><br>
 * You've been doing this long enough to know that to restore snow operations,
 * you need to check all fifty stars by December 25th.
 *
 * <p><br>
 * Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent
 * calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one
 * star. Good luck!
 *
 * <p><br>
 * You try to ask why they can't just use a weather machine ("not powerful enough") and where
 * they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot
 * of questions") and hang on did you just say the sky ("of course, where do you think snow comes
 * from") when you realize that the Elves are already loading you into a trebuchet ("please hold
 * still, we need to strap you in").
 *
 * <p><br>
 * As they're making the final adjustments, they discover that their calibration document
 * (your puzzle input) has been amended by a very young Elf who was apparently just excited
 * to show off her art skills. Consequently, the Elves are having trouble reading the values
 * on the document.
 *
 * <p><br>
 * The newly-improved calibration document consists of lines of text; each line originally
 * contained a specific calibration value that the Elves now need to recover. On each line,
 * the calibration value can be found by combining the first digit and the last digit
 * (in that order) to form a single two-digit number.
 *
 * <p><br>
 * For example:<br>
 * <ul>
 *     <li>
 *         1abc2
 *     </li>
 *     <li>
 *         pqr3stu8vwx
 *     </li>
 *     <li>
 *         a1b2c3d4e5f
 *     </li>
 *     <li>
 *         treb7uchet
 *     </li>
 * </ul>
 * In this example, the calibration values of these four lines are 12, 38, 15, and 77.
 * Adding these together produces 142.
 *
 * <p><br>
 * Consider your entire calibration document. What is the sum of all the calibration values?
 *
 * <p><br>
 * Your puzzle answer was 54338.
 *
 * <p><br>
 * --- Part Two ---<br><br>
 * Your calculation isn't quite right. It looks like some of the digits are actually spelled out w
 * ith letters: one, two, three, four, five, six, seven, eight, and nine also count
 * as valid "digits".
 *
 * <p><br>
 * Equipped with this new information, you now need to find the real first and last
 * digit on each line.
 *
 * <p><br>
 * For example:
 * <ul>
 *     <li>
 *         two1nine
 *     </li>
 *     <li>
 *         eightwothree
 *     </li>
 *     <li>
 *         abcone2threexyz
 *     </li>
 *     <li>
 *         xtwone3four
 *     </li>
 *     <li>
 *         4nineeightseven2
 *     </li>
 *     <li>
 *         zoneight234
 *     </li>
 *     <li>
 *         7pqrstsixteen
 *     </li>
 * </ul>
 * In this example, the calibration values are 29, 83, 13, 24, 42, 14, and 76. Adding
 * these together produces 281.
 *
 * <p><br>
 * What is the sum of all the calibration values?
 *
 * <p><br>
 * Your puzzle answer was 53389.
 */
public class Day01 implements AdventOfCodeSolution {
  private static final Logger LOGGER = LoggerFactory.getLogger(Day01.class);

  /**
   * Instantiates the solution instance.
   */
  public Day01() {

  }

  @Override
  public String getDefaultInputFilename() {
    return "inputs/input_day_01-01.txt";
  }

  @Override
  public void runPart1(List<String> input) {
    var start = Instant.now();
    var sum = getSumOfCalibrationValues(input, false);
    var end = Instant.now();

    LOGGER.info("The sum of the calibration values is: {}", sum);
    logTimings(LOGGER, start, end);
  }

  @Override
  public void runPart2(List<String> input) {
    var start = Instant.now();
    var sum = getSumOfCalibrationValues(input, true);
    var end = Instant.now();

    LOGGER.info("The sum of the real calibration values is: {}", sum);
    logTimings(LOGGER, start, end);
  }

  /**
   * Returns the sum of the calibration values specified in the List of input Strings.
   *
   * @param input The List of String calibration values.
   * @param real  If true, then words that spell digits will be interpreted as digits.
   * @return The sum of the calibration values specified in the List of input Strings.
   */
  public int getSumOfCalibrationValues(List<String> input, boolean real) {
    var instance = new Trebuchet(input);

    var values = instance.getCalibrationValues(real);

    return values.stream().reduce(0, Integer::sum);
  }
}

File: aoc-2017/src/main/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day01.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.string.InverseCaptcha
import java.time.Instant
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * **--- Day 1: Inverse Captcha ---**
 *
 * You're standing in a room with "digitization quarantine" written in LEDs along one wall.
 * The only door is locked, but it includes a small interface.
 * "Restricted Area - Strictly No Digitized Users Allowed."
 *
 * It goes on to explain that you may only leave by solving a captcha to prove
 * you're **not** a human. Apparently, you only get one millisecond to solve the
 * captcha: too fast for a normal human, but it feels like hours to you.
 *
 * The captcha requires you to review a sequence of digits (your puzzle input) and
 * find the **sum** of all digits that match the **next** digit in the list. The list is
 * circular, so the digit after the last digit is the **first** digit in the list.
 *
 * For example:
 *
 * - 1122 produces a sum of 3 (1 + 2) because the first digit (1) matches the second
 * digit and the third digit (2) matches the fourth digit.
 * - 1111 produces 4 because each digit (all 1) matches the next.
 * - 1234 produces 0 because no digit matches the next.
 * - 91212129 produces 9 because the only digit that matches the next one is the last digit, 9.
 *
 * **What is the solution** to your captcha?
 *
 * Your puzzle answer was **1102**.
 *
 * **--- Part Two ---**
 *
 * You notice a progress bar that jumps to 50% completion. Apparently, the door isn't
 * yet satisfied, but it did emit a star as encouragement. The instructions change:
 *
 * Now, instead of considering the **next** digit, it wants you to consider the digit
 * **halfway around** the circular list. That is, if your list contains 10 items, only
 * include a digit in your sum if the digit 10/2 = 5 steps forward matches it.
 * Fortunately, your list has an even number of elements.
 *
 * For example:
 *
 * - 1212 produces 6: the list contains 4 items, and all four digits match the digit 2 items ahead.
 * - 1221 produces 0, because every comparison is between a 1 and a 2.
 * - 123425 produces 4, because both 2s match each other, but no other digit has a match.
 * - 123123 produces 12.
 * - 12131415 produces 4.
 *
 * **What is the solution** to your new captcha?
 */
class Day01: AdventOfCodeSolution {
  companion object {
    private val log: Logger = LoggerFactory.getLogger(Day01::class.java)
  }

  override fun getDefaultInputFilename(): String = "inputs/input_day_01-01.txt"

  override fun runPart1(input: MutableList<String>) {
    val start = Instant.now()
    val answer = getCaptchaSolution(input.first())
    val end = Instant.now()

    log.info("$answer is the solution to the captcha problem!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: MutableList<String>) {
    val start = Instant.now()
    val answer = getCaptchaSolution(input.first(), true)
    val end = Instant.now()

    log.info("$answer is the solution to the captcha problem when using halfway!")
    logTimings(log, start, end)
  }

  /**
   * Solves the specified captcha and returns an integer that is the solution.
   *
   * @param input The numerical string captcha.
   * @param halfway If true, then digits are compared halfway around the circular list.
   * @return An integer that is the solution.
   */
  fun getCaptchaSolution(input: String, halfway: Boolean = false): Int {
    val instance = InverseCaptcha(input)
    return instance.solve(halfway)
  }
}
